File: src/openagents/tool/mcp_tool.py
import json
import logging
from contextlib import asynccontextmanager, AsyncExitStack

import httpx
from mcp import ClientSession, StdioServerParameters
from mcp.client.sse import sse_client
from mcp.client.stdio import stdio_client
from mcp.client.streamable_http import streamable_http_client
from mcp.types import TextContent, Tool, ListToolsResult
from pydantic import BaseModel, ConfigDict

from openagents.repository import mcp_server_repository


# MCP数据结构，session 与 tool_dict 为第三方任意类型
class McpServerInfo(BaseModel):
    model_config = ConfigDict(arbitrary_types_allowed=True)

    description: str
    session: ClientSession
    tool_dict: dict[str, Tool]


# 存储所有的MCP信息
MCP_DICT: dict[str, McpServerInfo] = {}


@asynccontextmanager
async def _register_mcp_client(name: str, description: str, proto_type: str, arg_dict: dict):
    # 创建客户端
    if proto_type == "streamable_http":
        http_client = httpx.AsyncClient(headers=arg_dict.get("headers"), timeout=httpx.Timeout(30.0, read=300.0))
        client = streamable_http_client(arg_dict["url"], http_client=http_client)
    elif proto_type == "sse":
        client = sse_client(arg_dict["url"], arg_dict.get("headers"))
    elif proto_type == "stdio":
        client = stdio_client(StdioServerParameters(command=arg_dict["command"], args=arg_dict["args"]))
    else:
        raise ValueError(f"Unknown proto type: {proto_type}")
    # 建立连接
    async with client as streams:
        read, write = streams[:2]
        async with ClientSession(read, write) as session:
            _ = await session.initialize()
            # 获取工具列表
            list_tools_result: ListToolsResult = await session.list_tools()
            MCP_DICT[name] = McpServerInfo(description=description, session=session, tool_dict={tool.name: tool for tool in list_tools_result.tools})
            logging.info(f"MCP client {name} started, having {len(list_tools_result.tools)} tools")
            # 等待
            yield
            # 结束
            MCP_DICT.pop(name)
            logging.info(f"MCP client {name} stopped")


@asynccontextmanager
async def lifespan():
    mcp_servers = await mcp_server_repository.list_mcp_servers()
    mcp_clients = [_register_mcp_client(server.name, server.description, server.type, server.model_dump(mode="json", exclude_none=True)) for server in mcp_servers]
    async with AsyncExitStack() as stack:
        for client in mcp_clients:
            try:
                await stack.enter_async_context(client)
            except Exception as e:
                if hasattr(e, "exceptions"):
                    logging.error(f"Error registering mcp client: {e.exceptions}")
                else:
                    logging.error(f"Error registering mcp client: {e}")
        yield


# 执行
async def execute(cmd_and_args: list[str], work_dir: str) -> tuple[str, bool]:
    # 1. mcp server list
    if len(cmd_and_args) == 3 and cmd_and_args[0] == "mcp" and cmd_and_args[1] == "server" and cmd_and_args[2] == "list":
        result = [{"name": name, "description": mcp_server_info.description} for name, mcp_server_info in MCP_DICT.items()]
        return json.dumps(result, ensure_ascii=False), False
    # 2. mcp server <server_name> tool list
    elif len(cmd_and_args) == 5 and cmd_and_args[0] == "mcp" and cmd_and_args[1] == "server" and cmd_and_args[3] == "tool" and cmd_and_args[4] == "list":
        server_name = cmd_and_args[2]
        if not server_name in MCP_DICT:
            return f"Unknown server {server_name}", True
        result = [{"name": tool.name, "description": tool.description} for tool in MCP_DICT[server_name].tool_dict.values()]
        return json.dumps(result, ensure_ascii=False), False
    # 3. mcp server <server_name> tool <tool_name> info
    elif len(cmd_and_args) == 6 and cmd_and_args[0] == "mcp" and cmd_and_args[1] == "server" and cmd_and_args[3] == "tool" and cmd_and_args[5] == "info":
        server_name, tool_name = cmd_and_args[2], cmd_and_args[4]
        if not server_name in MCP_DICT:
            return f"Unknown server {server_name}", True
        if not tool_name in MCP_DICT[server_name].tool_dict:
            return f"Unknown tool {tool_name}", True
        tool = MCP_DICT[server_name].tool_dict[tool_name]
        result = {"name": tool.name, "description": tool.description, "input_schema": tool.inputSchema}
        return json.dumps(result, ensure_ascii=False), False
    # 4. mcp server <server_name> tool <tool_name> call [tool_json_args]
    elif len(cmd_and_args) == 7 and cmd_and_args[0] == "mcp" and cmd_and_args[1] == "server" and cmd_and_args[3] == "tool" and cmd_and_args[5] == "call":
        server_name, tool_name, json_string = cmd_and_args[2], cmd_and_args[4], cmd_and_args[6]
        if not server_name in MCP_DICT:
            return f"Unknown server {server_name}", True
        if not tool_name in MCP_DICT[server_name].tool_dict:
            return f"Unknown tool {tool_name}", True
        session = MCP_DICT[server_name].session
        tool_result = await session.call_tool(tool_name, json.loads(json_string) if json_string else {})
        tool_content_list = [content.text if isinstance(content, TextContent) else content.type for content in tool_result.content]
        tool_content = tool_content_list[0] if len(tool_content_list) == 1 else json.dumps(tool_content_list, ensure_ascii=False)
        is_error = tool_result.isError
        return tool_content, is_error
    return "未知命令", True

File: src/tool/mcp_tool.rs
// MCP 客户端管理工具
use std::collections::HashMap;

use dashmap::DashMap;
use rmcp::model::{CallToolRequestParams, ClientCapabilities, ClientInfo, Implementation, Tool};
use rmcp::service::{Peer, RoleClient, RunningService};
use rmcp::transport::{ConfigureCommandExt, StreamableHttpClientTransport, TokioChildProcess};
use rmcp::ServiceExt;
use serde_json::Value;
use sqlx::SqlitePool;

use crate::repository::mcp_server_repository;

// MCP 服务信息
pub struct McpServerInfo {
    pub description: String,
    pub peer: Peer<RoleClient>,
    pub tool_dict: HashMap<String, Tool>,
}

// 存储所有的 MCP 信息
pub static MCP_DICT: std::sync::LazyLock<DashMap<String, McpServerInfo>> = std::sync::LazyLock::new(DashMap::new);

// 存储 RunningService 的句柄，用于保持连接
static MCP_HANDLES: std::sync::LazyLock<DashMap<String, tokio::task::JoinHandle<()>>> = std::sync::LazyLock::new(DashMap::new);

// 连接 MCP 服务并获取工具列表，返回 peer、工具表与保持连接的服务句柄
pub async fn connect_mcp_server(
    protocol_type: &str,
    url: Option<&str>,
    headers: Option<&Value>,
    command: Option<&str>,
    args: Option<&Value>,
) -> Result<(Peer<RoleClient>, HashMap<String, Tool>, RunningService<RoleClient, ClientInfo>), String> {
    // 创建客户端信息
    let client_info = ClientInfo::new(ClientCapabilities::default(), Implementation::new("openagents", env!("CARGO_PKG_VERSION")));

    // 按协议类型创建 transport
    let service = match protocol_type {
        "stdio" => {
            let command = command.ok_or_else(|| "missing command".to_string())?;
            let args: Vec<String> = match args {
                Some(Value::Array(arr)) => arr.iter().filter_map(|v| v.as_str().map(|s| s.to_string())).collect(),
                _ => Vec::new(),
            };
            let mut cmd = tokio::process::Command::new(command);
            for arg in &args {
                cmd.arg(arg);
            }
            // Windows 下隐藏子进程控制台窗口,避免桌面模式调用外部模型时弹出黑框
            #[cfg(windows)]
            {
                cmd.creation_flags(0x08000000);
            }
            let transport = TokioChildProcess::new(cmd.configure(|_c| {}))
                .map_err(|e| format!("failed to create stdio transport: {}", e))?;
            client_info
                .serve(transport)
                .await
                .map_err(|e| format!("failed to connect: {}", e))?
        }
        "streamable_http" => {
            let url = url.ok_or_else(|| "missing url".to_string())?;
            let mut config = rmcp::transport::streamable_http_client::StreamableHttpClientTransportConfig::with_uri(url);
            if let Some(Value::Object(headers_map)) = headers {
                let mut custom_headers = HashMap::new();
                for (k, v) in headers_map {
                    if let Some(val_str) = v.as_str() {
                        if let (Ok(header_name), Ok(header_value)) = (
                            k.parse::<http::HeaderName>(),
                            val_str.parse::<http::HeaderValue>(),
                        ) {
                            custom_headers.insert(header_name, header_value);
                        }
                    }
                }
                config = config.custom_headers(custom_headers);
            }
            let transport = StreamableHttpClientTransport::from_config(config);
            client_info
                .serve(transport)
                .await
                .map_err(|e| format!("failed to connect: {}", e))?
        }
        _ => return Err(format!("unsupported protocol type: {}", protocol_type)),
    };

    // 获取工具列表
    let peer = service.peer().clone();
    let tools = peer
        .list_all_tools()
        .await
        .map_err(|e| format!("failed to list tools: {}", e))?;
    let tool_dict: HashMap<String, Tool> = tools.into_iter().map(|t| (t.name.to_string(), t)).collect();

    Ok((peer, tool_dict, service))
}

// 初始化所有 MCP 客户端
pub async fn init_mcp_clients(db: &SqlitePool) {
    let servers = match mcp_server_repository::list_mcp_servers(db).await {
        Ok(s) => s,
        Err(e) => {
            tracing::error!("Failed to list MCP servers: {}", e);
            return;
        }
    };

    for server in servers {
        let name = server.name.clone();
        let description = server.description.clone();

        match connect_mcp_server(
            &server.protocol_type,
            server.url.as_deref(),
            server.headers.as_ref(),
            server.command.as_deref(),
            server.args.as_ref(),
        )
        .await
        {
            Ok((peer, tool_dict, service)) => {
                let count = tool_dict.len();
                MCP_DICT.insert(name.clone(), McpServerInfo { description, peer, tool_dict });
                tracing::info!("MCP client {} started, having {} tools", name, count);
                // 保持 service 运行
                let handle = tokio::spawn(async move {
                    let _ = service.waiting().await;
                });
                MCP_HANDLES.insert(name, handle);
            }
            Err(e) => {
                tracing::error!("Failed to start MCP client {}: {}", name, e);
            }
        }
    }
}

// 执行 MCP 命令
pub async fn execute(cmd_and_args: &[String]) -> (String, bool) {
    // 1. mcp server list
    if cmd_and_args.len() == 3 && cmd_and_args[0] == "mcp" && cmd_and_args[1] == "server" && cmd_and_args[2] == "list" {
        let result: Vec<serde_json::Value> = MCP_DICT
            .iter()
            .map(|entry| {
                serde_json::json!({
                    "name": entry.key(),
                    "description": entry.value().description,
                })
            })
            .collect();
        return (serde_json::to_string(&result).unwrap_or_default(), false);
    }
    // 2. mcp server <server_name> tool list
    else if cmd_and_args.len() == 5 && cmd_and_args[0] == "mcp" && cmd_and_args[1] == "server" && cmd_and_args[3] == "tool" && cmd_and_args[4] == "list" {
        let server_name = &cmd_and_args[2];
        let entry = match MCP_DICT.get(server_name) {
            Some(e) => e,
            None => return (format!("Unknown server {}", server_name), true),
        };
        let result: Vec<serde_json::Value> = entry
            .tool_dict
            .values()
            .map(|tool| {
                serde_json::json!({
                    "name": tool.name,
                    "description": tool.description,
                })
            })
            .collect();
        return (serde_json::to_string(&result).unwrap_or_default(), false);
    }
    // 3. mcp server <server_name> tool <tool_name> info
    else if cmd_and_args.len() == 6 && cmd_and_args[0] == "mcp" && cmd_and_args[1] == "server" && cmd_and_args[3] == "tool" && cmd_and_args[5] == "info" {
        let server_name = &cmd_and_args[2];
        let tool_name = &cmd_and_args[4];
        let entry = match MCP_DICT.get(server_name) {
            Some(e) => e,
            None => return (format!("Unknown server {}", server_name), true),
        };
        let tool = match entry.tool_dict.get(tool_name) {
            Some(t) => t,
            None => return (format!("Unknown tool {}", tool_name), true),
        };
        let result = serde_json::json!({
            "name": tool.name,
            "description": tool.description,
            "input_schema": tool.input_schema,
        });
        return (serde_json::to_string(&result).unwrap_or_default(), false);
    }
    // 4. mcp server <server_name> tool <tool_name> call [tool_json_args]
    else if cmd_and_args.len() == 7 && cmd_and_args[0] == "mcp" && cmd_and_args[1] == "server" && cmd_and_args[3] == "tool" && cmd_and_args[5] == "call" {
        let server_name = &cmd_and_args[2];
        let tool_name = &cmd_and_args[4];
        let json_string = &cmd_and_args[6];
        let entry = match MCP_DICT.get(server_name) {
            Some(e) => e,
            None => return (format!("Unknown server {}", server_name), true),
        };
        if !entry.tool_dict.contains_key(tool_name) {
            return (format!("Unknown tool {}", tool_name), true);
        }
        let arguments: Option<rmcp::model::JsonObject> = if json_string.is_empty() {
            None
        } else {
            match serde_json::from_str::<serde_json::Map<String, Value>>(json_string) {
                Ok(map) => Some(map),
                Err(e) => return (format!("Invalid JSON arguments: {}", e), true),
            }
        };
        let mut params = CallToolRequestParams::new(tool_name.clone());
        if let Some(args) = arguments {
            params = params.with_arguments(args);
        }
        match entry.peer.call_tool(params).await {
            Ok(tool_result) => {
                let tool_content_list: Vec<String> = tool_result
                    .content
                    .iter()
                    .map(|content| match content {
                        rmcp::model::ContentBlock::Text(text) => text.text.clone(),
                        other => format!("{:?}", other),
                    })
                    .collect();
                let tool_content = if tool_content_list.len() == 1 {
                    tool_content_list.into_iter().next().unwrap_or_default()
                } else {
                    serde_json::to_string(&tool_content_list).unwrap_or_default()
                };
                let is_error = tool_result.is_error.unwrap_or(false);
                return (tool_content, is_error);
            }
            Err(e) => {
                return (format!("Tool call failed: {}", e), true);
            }
        }
    }
    ("Unknown command".to_string(), true)
}
